src/resolver.rs

```diff
@@ -202,3 +281,93 @@ impl ResolverMap {
         }
     }
 }
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::mem::transmute;
+
+    use test_log::test;
+
+    use crate::elf::ElfParser;
+    use crate::mmap::Mmap;
+    use crate::zip;
+    use crate::BlazeSymbolizer;
+    use crate::SymbolType;
+
+
+    /// Check that we can create a path to an ELF inside an APK as expected.
+    #[test]
+    fn elf_apk_path_creation() {
+        let apk = Path::new("/root/test.apk");
+        let elf = Path::new("subdir/libc.so");
+        let path = create_apk_elf_path(apk, elf).unwrap();
+        assert_eq!(path, Path::new("/root/test.apk!/subdir/libc.so"));
+    }
+
+    /// Check that we can symbolize an address using DWARF.
+    #[test]
+    fn symbolize_apk() {
+        let test_apk = Path::new(&env!("CARGO_MANIFEST_DIR"))
+            .join("data")
+            .join("test.zip");
+
+        let mmap = Mmap::builder().exec().open(test_apk).unwrap();
+        let mmap = Rc::new(mmap);
+        let archive = zip::Archive::with_mmap(mmap.clone()).unwrap();
+        let so = archive
+            .entries()
+            .find_map(|entry| {
+                let entry = entry.unwrap();
+                (entry.path == Path::new("libtest-so.so")).then_some(entry)
+            })
+            .unwrap();
+
+        // We found the ELF shared object inside the archive. Now look
+        // up the address of the `the_answer` function inside of it.
+        let elf_parser = ElfParser::from_mmap(mmap.clone(), so.data_offset).unwrap();
+        let opts = FindAddrOpts {
+            sym_type: SymbolType::Function,
+            ..Default::default()
+        };
+        let symbols = elf_parser.find_address("the_answer", &opts).unwrap();
+        // There is only one symbol with this address in the shared
+        // object.
+        assert_eq!(symbols.len(), 1);
+        let symbol = symbols.first().unwrap();
+
+        let the_answer_addr = unsafe {
+            mmap.as_ptr()
+                .add(so.data_offset)
+                // The address as reported by ELF is just an offset for
+                // our intents and purposes, because the symbol is
+                // relative to the beginning of the file.
```

Comment:
dangerous comment, this is not true in general, remember that virtual address offset is not equal to file offset as a general rule

Reply:
It's an explanation of what the test code does, not a general remark about all things ELF.